File: problems/ruby/160721-single-number/code.rb
# https://leetcode.com/problems/single-number/
#
# @param {Integer[]} nums
# @return {Integer}

def single_number(nums)
  # There is a trick with XOR: x ^ x = 0 and x ^ 0 = x
  nums.reduce { |memo, value| memo ^ value }
end

puts single_number([2, 2, 1]).to_s
puts single_number([4, 1, 2, 1, 2]).to_s

File: problems/javascript/16072021-single-number/code.js
Array.prototype.count = function (val) {
  return this.filter((elem) => elem === val).length
}

/**
 * https://leetcode.com/problems/single-number/
 *
 * @param {number[]} nums
 * @return {number}
 */

const singleNumber = (nums) => {
  for (let i = 0; i < nums.length; i += 1) {
    if (nums.count(nums[i]) === 1) return nums[i]
  }
}

/**
 * Better solution using XOR trick
 * x ^ x = 0
 * x ^ 0 = x
 */

const singleNumberBetter = (nums) => {
  return nums.reduce((memo, value) => memo ^ value)
}

console.log(singleNumber([2, 2, 1]))
console.log(singleNumber([4, 1, 2, 1, 2]))
console.log(singleNumberBetter([2, 2, 1]))
console.log(singleNumberBetter([4, 1, 2, 1, 2]))
